File: special/fireups/reverse.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

void reverse_str(char* begin)
{
    char* end = begin + strlen(begin) - 1;
    char swp;
    while (begin < end)
    {
        swp = *end;
        *end-- = *begin;
        *begin++ = swp;
    }
}

int main(int argc, char* argv[])
{
    char* forward = argv[1];
    char* reverse = strdup(forward);
    if (reverse != NULL)
    {
        reverse_str(reverse);

        printf("forward: %s\n", forward);
        printf("reverse: %s\n", reverse);

        free(reverse);
    }
    else
        printf("reversing of string failed due to insufficient memory.\n");

    return 0;
}

File: special/fireups/reverse.py
from sys import argv

def reverse_str_c_style(cstr):
    begin, end = 0, len(cstr) - 1
    ret_list = list(cstr)
    while begin < end:
        ret_list[begin], ret_list[end] = ret_list[end], ret_list[begin]
        begin += 1
        end -= 1
    return "".join(ret_list)

forward, reverse = argv[1], reverse_str_c_style(argv[1])
print "forward: %s" % forward
print "reverse: %s" % reverse
# print "reverse: %s" % "".join(list(reversed(forward)))
